src/gretel_synthetics/generate_parallel.py

```diff
@@ -0,0 +1,185 @@
+import multiprocessing
+from dataclasses import dataclass, field
+from typing import Iterable, List, Optional, Union, Tuple
+import queue
+import cloudpickle
+import sys
+import os
+
+from .generator import Generator, Settings, gen_text
+
+mp = multiprocessing.get_context('spawn')  # fork does not work with tensorflow
+
+
+@dataclass
+class _WorkerStatus:
+    """
+    Status of a parallel worker to be communicated back to the main thread after each chunk.
+    """
+
+    lines: List[gen_text] = field(default_factory=list)
+    """Generated lines in this chunk, if any. This contains both valid and invalid lines."""
+
+    exception: Optional[BaseException] = None
+    """Exception that was encountered during processing. Implies done=True when not None."""
+
+    done: bool = False
+    """Flag indicating whether the worker is complete."""
+
+    def serialize(self) -> bytes:
+        return cloudpickle.dumps(self)
+
+    @staticmethod
+    def deserialize(serialized: bytes) -> "_WorkerStatus":
+        obj = cloudpickle.loads(serialized)
+        if not isinstance(obj, _WorkerStatus):
+            raise TypeError("serialized object is of type {}, not WorkerStatus".format(type(obj).__name__))
+        return obj
+
+
+def split_work(parallelism: Union[int, float], total_lines: int, chunk_size: int = 5) -> Tuple[int, List[int]]:
+    """
+    Given a parallelism setting and a number of total lines, split the work across workers.
+
+    Args:
+        parallelism: The number of concurrent workers to use. ``1`` (the default) disables parallelization,
+            while a non-positive value means "number of CPUs + x" (i.e., use ``0`` for using as many workers
+            as there are CPUs). A floating-point value is interpreted as a fraction of the available CPUs,
+            rounded down.
+        total_lines: The total number of lines to generate.
+        chunk_size: the size of an individual unit of work to be distributed to workers.
+
+    Returns:
+        A pair consisting of the number of required workers, and the list of chunks of work.
+    """
+    num_chunks = (total_lines - 1) // chunk_size + 1
+    if isinstance(parallelism, float):
+        num_workers = int(mp.cpu_count() * parallelism)
+    elif parallelism <= 0:
+        num_workers = mp.cpu_count() + parallelism
+    else:
+        num_workers = parallelism
```

Comment:
If we're going to provide this level of inference on `parallelism` I just wanted to verify that providing a negative floating value is intended to set the CPU count to 1 based on the `max(num_workers, 1)` logic.

On my local machine with 12 avail CPUs, .5 gives me 6 and -.5 gives me 1.

Reply:
Would be good to have a unit test for this, too to validate the permutations of inputs with expected outputs